File: addBinary.py
def addBinary(a,b):                     # TC O(max(n,m)) // SC O(max(n,m))
    a,b = a[::-1], b[::-1]
    result=""
    carry=0

    for i in range(max(len(a),len(b))):
        digitA=int(a[i]) if i < len(a) else 0
        digitB=int(b[i]) if i < len(b) else 0

        total= digitA + digitB + carry
        char = str(total % 2)     #because we want binary that is base 2 
        result = char + result    #we are adding strings here, so order of addition(appending)matters

        carry=total // 2          # taking care of carry

    if carry:
        result= "1"+ result
    return result


#Approach- Built in functions 

def addBinary1(a,b):          # TC O(n+m)
    c = bin(int(a,2) + int(b,2))
    return c[2:]


#Approach- Bit manipulation 

def addBinary2(a,b):               # TC O(n+m) // SC O(max(n,m))
     x, y = int(a, 2), int(b, 2)
     while y:
         answer = x ^ y            # XOR ( ^ )returns 1 when bits are different else 0
         carry = (x & y) << 1      # AND ( & )returns 1 when both bits are 1 else 0
         x, y = answer, carry      # << 1 --> leftshift bit by 1
         #x,y= x ^ y, (x & y) <<1
     return bin(x)[2:]

a=input("Enter first binary string: ")
b=input("Enter second binary string: ")
print(addBinary(a,b))
print(addBinary1(a,b))
print(addBinary2(a,b))

File: addBinary.java
// Approach 1- Bit by bit computation
class Solution { // TC O(max(n,m) SC O(max(n,m)
  public String addBinary(String a, String b) {
    int n = a.length(), m = b.length();
    if (n < m) return addBinary(b, a);
    int L = Math.max(n, m);

    StringBuilder sb = new StringBuilder();
    int carry = 0, j = m - 1;
    for(int i = L - 1; i > -1; --i) {
      if (a.charAt(i) == '1') ++carry;
      if (j > -1 && b.charAt(j--) == '1') ++carry;

      if (carry % 2 == 1) sb.append('1');
      else sb.append('0');

      carry /= 2;
    }
    if (carry == 1) sb.append('1');
    sb.reverse();

    return sb.toString();
  }
}

// Approach 2- Bit manipulation 
import java.math.BigInteger;
class Solution { // TC O(m+n) SC O(max(n,m)
  public String addBinary(String a, String b) {
    BigInteger x = new BigInteger(a, 2);
    BigInteger y = new BigInteger(b, 2);
    BigInteger zero = new BigInteger("0", 2);
    BigInteger carry, answer;
    while (y.compareTo(zero) != 0) {
      answer = x.xor(y);
      carry = x.and(y).shiftLeft(1);
      x = answer;
      y = carry;
    }
    return x.toString(2);
  }
}

class addBinary{
  public static void main(String[] args) {
    String a="11";
    String b="1";
    Solution s= new Solution();
    System.out.println(s.addBinary(a,b));
  }
}